src/coreclr/jit/importer.cpp

```diff
@@ -16996,19 +16995,29 @@ void Compiler::impImportBlockCode(BasicBlock* block)
 
                 op3 = impPopStack().val; // Size
                 op2 = impPopStack().val; // Value
-                op1 = impPopStack().val; // Dest
+                op1 = impPopStack().val; // Dst addr
 
                 if (op3->IsCnsIntOrI())
                 {
                     size = (unsigned)op3->AsIntConCommon()->IconValue();
                     op1  = new (this, GT_BLK) GenTreeBlk(GT_BLK, TYP_STRUCT, op1, typGetBlkLayout(size));
+                    op1  = gtNewBlkOpNode(op1, op2, (prefixFlags & PREFIX_VOLATILE) != 0, false);
                 }
                 else
                 {
-                    op1  = new (this, GT_DYN_BLK) GenTreeDynBlk(op1, op3);
+                    if (!op2->IsIntegralConst(0))
+                    {
+                        op2 = gtNewOperNode(GT_INIT_VAL, TYP_INT, op2);
+                    }
+
+                    op1  = new (this, GT_STORE_DYN_BLK) GenTreeStoreDynBlk(op1, op2, op3);
                     size = 0;
+
+                    if ((prefixFlags & PREFIX_VOLATILE) != 0)
+                    {
+                        op1->gtFlags |= GTF_BLK_VOLATILE;
+                    }
                 }
-                op1 = gtNewBlkOpNode(op1, op2, (prefixFlags & PREFIX_VOLATILE) != 0, false);
```

Comment:
Yes. `gtBlockOpInit`:
1) Checks for self-assignment to not confuse liveness. Doesn't matter for `STORE_DYN_BLK` because any local possibly defined by it will be address-exposed.
2) Propagates side effects flags. `StoreDynBlk`'s constructor already does that.
3) Sets `GTF_BLK_VOLATILE`. We do it here manually.
4) Labels assignments to SIMD locals so that later we know not try to promote them (optional optimization). Doesn't matter for `STORE_DYN_BLK` because, as above, any locals possibly defined by it will be address-exposed and thus not subject for enregistration anyway.